src/libsyntax/parse/unescape.rs

```diff
@@ -0,0 +1,512 @@
+        let res = scan_escape(first_char, &mut chars, Mode::Byte)?;
+        if chars.next().is_some() {
+            return Err(EscapeError::MoreThanOneChar);
+        }
+        Ok(byte_from_char(res))
+    })();
+    res.map_err(|err| (literal_text.len() - chars.as_str().len(), err))
+}
+
+/// Takes a contents of a string literal (without quotes) and produces a
+/// sequence of escaped characters or errors.
+pub(crate) fn unescape_byte_str<F>(literal_text: &str, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<u8, EscapeError>),
+{
+    unescape_str_or_byte_str(literal_text, Mode::ByteStr, &mut |range, char| {
+        callback(range, char.map(byte_from_char))
+    })
+}
+
+fn byte_from_char(c: char) -> u8 {
+    let res = c as u32;
+    assert!(res <= u8::max_value() as u32, "guaranteed because of Mode::Byte");
+    res as u8
+}
+
+#[derive(Clone, Copy)]
+enum Mode {
+    Char,
+    Str,
+    Byte,
+    #[allow(unused)]
+    ByteStr,
+}
+
+impl Mode {
+    fn in_single_quotes(self) -> bool {
+        match self {
+            Mode::Char | Mode::Byte => true,
+            Mode::Str | Mode::ByteStr => false,
+        }
+    }
+
+    fn in_double_quotes(self) -> bool {
+        !self.in_single_quotes()
+    }
+
+    fn allow_unicode(self) -> bool {
+        match self {
+            Mode::Byte | Mode::ByteStr => false,
+            Mode::Char | Mode::Str => true,
+        }
+    }
+}
+
+fn scan_escape(first_char: char, chars: &mut Chars<'_>, mode: Mode) -> Result<char, EscapeError> {
+    if first_char != '\\' {
+        return match first_char {
+            '\t' | '\n' => Err(EscapeError::EscapeOnlyChar),
+            '\r' => Err(if chars.clone().next() == Some('\n') {
+                EscapeError::EscapeOnlyChar
+            } else {
+                EscapeError::BareCarriageReturn
+            }),
+            '\'' if mode.in_single_quotes() => Err(EscapeError::EscapeOnlyChar),
+            '"' if mode.in_double_quotes() => Err(EscapeError::EscapeOnlyChar),
+            _ => {
+                if !mode.allow_unicode() && !first_char.is_ascii() {
+                    return Err(EscapeError::NonAsciiCharInByte);
+                }
+                Ok(first_char)
+            }
+        };
+    }
+
+    let second_char = chars.next().ok_or(EscapeError::LoneSlash)?;
+
+    let res = match second_char {
+        '"' => '"',
+        'n' => '\n',
+        'r' => '\r',
+        't' => '\t',
+        '\\' => '\\',
+        '\'' => '\'',
+        '0' => '\0',
+
+        'x' => {
+            let hi = chars.next().ok_or(EscapeError::TooShortHexEscape)?;
+            let hi = hi.to_digit(16).ok_or(EscapeError::InvalidCharInHexEscape)?;
+
+            let lo = chars.next().ok_or(EscapeError::TooShortHexEscape)?;
+            let lo = lo.to_digit(16).ok_or(EscapeError::InvalidCharInHexEscape)?;
+
+            let value = hi.checked_mul(16).unwrap().checked_add(lo).unwrap();
+
+            if mode.allow_unicode() && value > 0x7F {
```

Comment:
I'm not sure why is the `allow_unicode` condition necessary, "large" ASCII escapes are always prohibited.

Also nit:
```rust
let value = (hi * 16 + lo) as char;
if !value.is_ascii() {
    error
}
value
```

Reply:
>large" ASCII escapes are always prohibited.

They are allowed in bytes: 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e6beaacbbda984e84f8ae1f4ca85158d

Reply:
>Also nit:

That won't work unfortunately:  `u16 as char` is not a thing due to surrogates, and ther's no `is_ascii` on `u16` (perhaps there should be?). Doing checked conversion to char/u8 and then checking range again doesn't seem better. Though I guess adding `is_ascii(u16)` would make this more readable

Reply:
I see.